Source/DafnyCore/AST/Attributes.cs

```diff
@@ -207,6 +233,232 @@ public static bool ContainsMatchingValue(Attributes attrs, string nm, ref object
+    if (Get(bindings, 0, out var expr) && expr != null) {
+      return A(name, expr);
+    }
+
+    return A(name);
+  }
+
+  // Given a user-supplied @-attribute, expand it if recognized as builtin to an old-style attribute
+  // or mark i as not built-in for later resolution
+  public static Attributes ExpandAtAttribute(Program program, UserSuppliedAtAttribute atAttribute) {
+    var toMatch = atAttribute.Arg;
+    var name = atAttribute.UserSuppliedName;
+    var bindings = atAttribute.UserSuppliedPreResolveBindings;
+
+    if (name == null) {
+      program.Reporter.Error(MessageSource.Resolver, atAttribute.RangeToken, "@-Attribute not recognized: " + atAttribute.ToString());
+      return null;
+    }
+
+    if (!TryGetBuiltinAtAttribute(name, out var builtinSyntax) || builtinSyntax == null) {
+      atAttribute.Builtin = false; // Will be resolved after
+      return null;
+    }
+
+    if (name != "Induction" && name != "Trigger") {
+      var formals = builtinSyntax.Args.Select(arg => arg.ToFormal()).ToArray();
+      ResolveLikeDatatypeConstructor(program, formals, name, toMatch.tok, bindings);
+    } // For @Induction and @Trigger, resolution is done in the generated version of the attributes
+
+    atAttribute.Builtin = true;
+    atAttribute.Arg.Type = Type.Int; // Dummy type to avoid crashes
+
+    switch (name) {
+      case "Compile": {
+          return A1("compile", bindings);
+        }
+      case "Fuel": {
+          if (Get(bindings, 0, out var lowFuel) && lowFuel != null) {
+            if (Get(bindings, 1, out var highFuel) && highFuel != null) {
+              if (Get(bindings, 2, out var functionName) && IsStringNotEmpty(functionName)) {
+                return A("fuel", functionName, lowFuel, highFuel);
+              }
+
+              return A("fuel", lowFuel, highFuel);
+            }
+
+            return A("fuel", lowFuel);
+          }
+
+          return A("fuel");
+        }
+      case "Induction": {
+          return A("induction", bindings.ArgumentBindings.Select(binding => binding.Actual).ToArray());
+        }
+      case "IsolateAssertions": {
+          return A("isolate_assertions");
+        }
+      case "Options": {
+          return A1("options", bindings);
+        }
+      default: {
+          throw new Exception("@-Attribute added to Attributes.BuiltinAtAttributes needs to be handled here");
+        }
+    }
+  }
+
+  // List of built-in @-attributes with their definitions.
+  // This list could be obtained from parsing and resolving a .Dfy file
+  // but for now it's good enough.
+  public static readonly List<BuiltInAtAttributeSyntax> BuiltinAtAttributes = new() {
+    B("Compile").WithArg(TupleItem0Name, Type.Bool, DefaultBool(true)),
+    B("Fuel")
+      .WithArg("low", Type.Int, DefaultInt(1))
+      .WithArg("high", Type.Int, DefaultInt(2))
+      .WithArg("functionName", Type.ResolvedString(), DefaultString("")),
+    B("Induction"), // Resolution is different
```

Comment:
I would leave this one out in the first batch, for two reasons:
1. I don't think this should be a single datatype constructor, I imagine it as something like `datatype Induction = NoInduction | InductionByAllVariables | InductionBy(???)`
2. I don't know how to declare a list of bound variables as an argument to a datatype constructor.

We may not ever want to support this as an @-attribute.

Reply:
Same goes for `@Trigger`

Reply:
I like the rule that attribute arguments must be resolve-able without knowing which attribute it is, so purely based on the location in which the attribute occurs, which would likely mean `@Trigger` can not be an attribute.

Reply:
I included `@Trigger` in the first batch _because_ I want us to have a discussion and clear agreement about it before I continue, so thanks for bringing this to me.

As of today, `{:induction]` and `{:trigger}` are the only attributes which can't reasonably considered datatype constructors, because although their input needs to be resolved, they can have any number of arguments, their arguments can be of any type, and the arguments are not used for their literal values but for names and patterns.

`{:Induction}` takes a set of variables in scope, like the `assigned()` ghost construct. It won't accept any expression, but it can accept an unbounded number of variables. We can't enforce this in the Dafny language itself as you remarked in your 2., @robin-aws .
Similarly, `{:trigger}` has to take patterns with variables that are in the scope, and all its values must have a type but it could be whatever type.

As a recall, the goals when having new attributes were the following
1. Make sure our users don't mistype a name (e.g. type `{:triggers}` instead of `{:trigger}`)
2. Make sure arguments to attributes have an expected type
3. Open up the ability for user to define their own typed attributes.
4. **Make it possible to migrate entirely** to the new @-attribute syntax so that we don't have any more inline attributes, that are sometimes ambiguous where to place (like `{:trigger}`) only prefix attributes.

That attributes be datatypes constructors felt the right choice as illustrated that we chose the `PascalCase`, regular constructor call syntax and even parameter naming when appropriate. I had then searched for ways to encode the special attributes `{:induction}` and `{:trigger}` but found none reasonable given the special nature of these constructors.
Hence, for built-in attributes, the idea `attribute = datatype constructor` gives no reasonable way to satisfy goal 4.
Fortunately, I've found that relaxing that initial idea, i.e. not resolving  `@Induction` and `@Trigger` as datatypes constructors but only their name and their arguments, given their nature, makes it possible to satisfy all four goals at once. Since these two attributes have always been exceptions, these PRs is not changing this when including them in the new style syntax.

Moreover, having now implemented Robin's request [here](https://github.com/dafny-lang/dafny/pull/5825#discussion_r1797387446), it even feels more ok to me now that built-in attributes can be more flexible than datatypes constructors at least in two cases, as they are not reported as such.

Note that errors on user-defined attributes will still be reported as "datatype constructors", as their definition will be in scope.

> I like the rule that attribute arguments must be resolve-able without knowing which attribute it is, so purely based on the location in which the attribute occurs, which would likely mean @Trigger can not be an attribute.

The upcoming resolution of the `@Trigger` attribute does resolve attribute arguments purely based on the location in which the attribute occurs, thanks to the rewriting phase. Would you like to explain a bit more your thoughts? I'd love to understand better.

Reply:
> Make it possible to migrate entirely to the new @-attribute syntax so that we don't have any more inline attributes, that are sometimes ambiguous where to place (like {:trigger}) only prefix attributes.

Moving away from old-style attributes doesn't mean everything they supported needs to be supported by @-attributes, right? We can have custom syntax for :trigger and :induction.

I think it could cause unnecessary migration effort if we first release @Trigger and @Induction, and then later replace those with custom syntax.

> it even feels more ok to me now that built-in attributes can be more flexible than datatypes constructors at least in two cases

It feels wrong to me. If it doesn't behave like all the other attributes, why is it an attribute to begin with?

Reply:
> I think that, as a user of a programming language, I would expect any new language feature that seems to aim at replacing another language feature to give me more, if not as many, ways to express what was possible to express before, but not less.

I don't see how this applies to our situation, since @ attributes do not need to aim to replace all `{:` attributes